### PR TITLE
New feature:  multiple subscriptions per customer.

### PR DESCRIPTION
**Pull request by h-c-c**
and some fixes for a couple other problems.  I use it in production.  NB: You'll be super sad if you try an use this on an existing production install without looking at UPGRADE.txt.

### CODE REVIEW

**Comment by BorislavZlatanov** [comment]
Very happy to see subscription_id in place of invoice_id as a way to track subscriptions. Kudoz!

Some notes:
1) In stripe.php, on line 615 there is `$stripe_response = $stripe_customer->subscriptions->create($cust_sub_params);` and after that there is no check whether the subscription was successfully created. I'm wondering what would happen if Stripe can't create the subscription because, let's say, a customer does not have sufficient funds in their credit card. According to the Stripe API (https://stripe.com/docs/api#create_subscription), they return an error if creation of subscription fails. Do you think it makes sense for us to output some error message to the user if that happens? 

2) When we use CRM_Core_DAO::executeQuery(), I think that we need to check not only if a result is returned but also if the data we expect is in the result (after doing a fetch). Otherwise the data may not be there and we run into trouble. For example:

```

if (!empty($rel_info_query)) {
  $rel_info_query->fetch();

  if (!empty($rel_info_query->invoice_id)) {
    $invoice_id = $rel_info_query->invoice_id;
  } else {
    CRM_Core_Error::Fatal("Error relating this customer ($customer_id) to the one in civicrm_stripe_subscriptions");
    exit();
  }
}

```

If we don't check $rel_info_query->invoice_id after doing a fetch, we don't know if invoice_id is in $rel_info_query. Can you do this sort of check in getOrigInvoiceId()?

3) In webhook.php, after line 101, `if (!empty($orig_contrib['values'][0]['contribution_id'])) {`, it seems that you treat the two conditions inside that if statement as two different cases. One case is when the subscription_id equals the trxn_id and the other case is when the contribution status is Pending. But I think that when a user signs up for a recurring contribution, then the contribution will have both the status set to Pending and the trxn_id equal to subscription_id (as this is how you set it up in Stripe.php). Those two conditions should really be one condition, IMO. A newly created recurring contribution would meet both conditions.

If you are running this code in production, my guess is that your users are not getting a receipt for recurring contributions. Either this or stripe is sending two webhooks. Because after trxn_id and fee are set in the first if statement, there is a return so we never get to contribution.completetransaction. 

Does this make sense? I may be missing something but I've spent quite a lot of hours banging my head on all of this. I have a suggestion for how to fix it but I'm curious to first hear whether you see it the same way. 

4) contribution.completetransaction is inserting the fee_amount properly, but you need CiviCRM 4.6.11. See the current code, line 101 and https://issues.civicrm.org/jira/browse/CRM-17734

**Comment by h-c-c** [comment]
Thanks Boris.  

Re 1.  I get "card declined". What do you get? 
screen shot 2016-02-14 at 3 06 19 pm   Try the card-declined number here:
https://stripe.com/docs/testing   Feel free to test it yourself.  ;-) 

Re: 2. Have you ever seen an entry in civicrm_stripe_subscriptions with a NULL invoice_id?  It seems that is the condition you're concerned with. Anything is possible, but I've never seen it.  If it is possible, sure let's add a check.   Looks like if forgot to return $end_time in that function.

Re: 3.  You're off in a couple different ways....the comment starting on line 105 was supposed to help clear this up. First, Webhook.php is not at all needed to complete a transaction.  You can easily test this yourself by removing Webhook.php or otherwise breaking it and make a contribution through an online contribution page.    You'll see that the contribution is Complete.  Generally, I think the reason for code that completes a transaction within Webhopok.php  is there not for initial payments of a subscription, but for recurring ones that fail for some reason, and then get sorted out later, where Stripe makes another attempt and it is successful.

> But I think that when a user signs up for a recurring contribution, then the contribution will have both the status set to Pending and the trxn_id equal to subscription_id (as this is how you set it up in Stripe.php). Those two conditions should really be one condition, IMO. A newly created recurring contribution would meet both conditions.

 Secondly, any pending contribution will never have a subscription id in the trxn_id field.  How can it?  Stripe only returns a subscription id if the card is good and the payment goes through.    These things that you think will happen at the same time are mutually exclusive.  I hope now it might become more clear why I'm temporarily putting subscription id into the trxn_id field.   Because the status is already 1, we need another way to store state info to ignore the first invoice.payment.succeeded.   

> If you are running this code in production, my guess is that your users are not getting a receipt for recurring contributions. Either this or stripe is sending two webhooks. Because after trxn_id and fee are set in the first if statement, there is a return so we never get to contribution.completetransaction. 

About email...It's funny, I was going to make a similar comment to you.  

> //Create a copy record of the original contribution and send out email receipt
> - $result = civicrm_api3('Contribution', 'repeattransaction', array(
> - 'sequential' => 1,
> - 'original_contribution_id' => $original_contribution['id'],
> - 'contribution_status_id' => "Completed",
> - 'trxn_id' => $transaction_id //Insert new transaction ID
> - //'invoice_id' => $new_invoice_id - contribution.repeattransaction
>   doesn't support it currently
> - ));

You comment talks about sending email, but then you leave out the argument that actually does it.  

> 'is_email_receipt' => 1
> Unless it happens somewhere else in the code that I missed.   

Re: 4.   Kudos on the patch!  But fyi, that didn't make it into 4.6.11.  We can certainly update this accordingly. 

Recurring payments go through fine with receipts....I hope this makes sense now.  Let me know!

**Comment by BorislavZlatanov** [comment]
Re: 1. Yeah, I tested now I also get "card declined". Cool. :)

Re: 2. Let's test this. When I put a non-existing customer_id in the query parameters so that I have:

```

$query_params = array(
  1 => array('blahblah', 'String'),
);

$rel_info_query = CRM_Core_DAO::executeQuery("SELECT invoice_id, end_time
  FROM civicrm_stripe_subscriptions
  WHERE customer_id = %1",
  $query_params);  

if (!empty($rel_info_query)) {
  $rel_info_query->fetch();
echo '<pre>';
var_dump($rel_info_query);
echo '</pre>';
  if(!empty($rel_info_query->invoice_id)) {
    $invoice_id = $rel_info_query->invoice_id;
    $end_time = $rel_info_query->end_time;
  } else {
    CRM_Core_Error::Fatal("Error relating this customer ($customer_id) to the one in civicrm_stripe_subscriptions");
    exit();
  }
}

```

Here's what I get from the var_dump above with a non-existing customer:

```

object(CRM_Core_DAO)#963 (12) {
  ["_DB_DataObject_version"]=>
  string(6) "1.8.12"
  ["__table"]=>
  NULL
  ["N"]=>
  int(0)
  ["_database_dsn"]=>
  string(0) ""
  ["_database_dsn_md5"]=>
  string(32) "e37733545fb90ac5bdbbc37234basrab"
  ["_database"]=>
  string(17) "mydb"
  ["_query"]=>
  array(7) {
    ["condition"]=>
    string(0) ""
    ["group_by"]=>
    string(0) ""
    ["order_by"]=>
    string(0) ""
    ["having"]=>
    string(0) ""
    ["limit_start"]=>
    string(0) ""
    ["limit_count"]=>
    string(0) ""
    ["data_select"]=>
    string(1) "*"
  }
  ["_DB_resultid"]=>
  int(8)
  ["_resultFields"]=>
  bool(false)
  ["_link_loaded"]=>
  bool(false)
  ["_join"]=>
  string(0) ""
  ["_lastError"]=>
  bool(false)
}

```

Here's what I get with an existing customer:

```

object(CRM_Core_DAO)#963 (14) {
  ["_DB_DataObject_version"]=>
  string(6) "1.8.12"
  ["__table"]=>
  NULL
  ["N"]=>
  int(1)
  ["_database_dsn"]=>
  string(0) ""
  ["_database_dsn_md5"]=>
  string(32) "e37733545fb90ac5bdbbc37234basrab"
  ["_database"]=>
  string(17) "mydb"
  ["_query"]=>
  bool(false)
  ["_DB_resultid"]=>
  int(8)
  ["_resultFields"]=>
  bool(false)
  ["_link_loaded"]=>
  bool(false)
  ["_join"]=>
  string(0) ""
  ["_lastError"]=>
  bool(false)
  ["invoice_id"]=>
  string(32) "5bhasn38bmuu66df281cfbec9a6ee"
  ["end_time"]=>
  string(1) "0"
}

```

Seems that $rel_info_query->invoice_id does not exist if the query doesn't match any customer.

Re: 3. I just tested this with your code. I made a recurring contribution without having any webhook set up in Stripe.com. And the contribution remains Pending in Civi. In Stripe.com the payment is completed, but in Civi the contribution is Pending. This is what happens on my end. So yeah, I am baffled.

Can you do the same test and see what happens? Make sure no webhook endpoint is set up on Stripe.com and submit a RECURRING  (not single) contribution. Then see what the contribution status is in Civi. Keep in mind, we are talking about the status in Civi, not in Stripe.com.

Btw, about `'is_email_receipt' => 1`, I imagine it defaults to true because it sends out email receipts even without providing that argument.

Re: 4. Hmm. I imagine the Fix Version/s in the Jira issue is incorrect. But anyways, I downloaded 4.6.12 and the patch is in there. 

Cheers, buddy. :)

**Comment by h-c-c** [comment]
Re: 2.  I don't see the relevance to my fork and this PR as I don't ever run a CRM_Core_DAO::executeQuery() with customer_id as a query parameter in Webhook.php.  I think I can safely assume you're not referring to Stripe.php since it does check before executing a query using customer_id, and you do mention the getOrigInvoiceId() function.   Is it possible that you're not working with my fork?  That might explain the different test results also.  I'm happy to comment on _my_ code.  ;-)  OK, pal?   

Re: 3.  Yes, of course. I meant do a recurring contribution without Webhook.php in place.  It would be non-nonsensical to do a single charge.  Single charges don't generate an invoice.payment.succeeded event at all.  That wouldn't satisfy the case that we're dealing with and just wouldn't be in the context of the conversation. 

> Can you do the same test and see what happens? Make sure no webhook endpoint is set up on Stripe.com and submit a RECURRING (not single) contribution. Then see what the contribution status is in Civi. Keep in mind, we are talking about the status in Civi, not in Stripe.com.

Do you know this test you're suggesting for me is functionally the same as removing Webhook.php?  Just to clarify:  When you do this test, in your civicrm_contributions table, the contribution_status_id is something other than 1?

You may have to increment the version number in info.xml to get civi to use this update.   I've been doing a 1.9.2.x convention.  Or try clearing your caches.  

> Btw, about 'is_email_receipt' => 1, I imagine it defaults to true because it sends out email receipts even without providing that argument.

Interesting to note, while the api contribution repeatransaction  does send an email for live contributions by default (you are correct!), it will refuse to send the email even if 'is_email_receipt' => 1 is explicitly set when attempting to repeat a test contribution.  

Re: 4.  Yes! I see it did make it into 4.6.12! Nice!

**Comment by BorislavZlatanov** [comment]
Re: 2. Ah, I could have expressed myself more concretely and used your code in my example. I wanted to show that when CRM_Core_DAO::executeQuery() returns no results, it still nevertheless returns an object. 

In getOrigInvoiceId() you are doing a query with subscription_id in the WHERE clause. If no results are found, then the check `if (!empty($sub_info_query))` would still evaluate to true. The query returns an object even if it finds no matches. So an additional check is needed after `$sub_info_query->fetch();`. This check could be for example whether an invoice_id was returned. In code:

```

if (!empty($sub_info_query)) {
      $sub_info_query->fetch();

      if(!empty($sub_info_query->invoice_id)) {
        $end_time = $sub_info_query->end_time;
        $original_invoice_id = $sub_info_query->invoice_id;
      }
    else {
      CRM_Core_Error::Fatal("Error relating this subscription id ($subscription_id) to the one in civicrm_stripe_subscriptions for customer with id ($customer_id) ");
      exit();
    }
}
    return $original_invoice_id;         

```

Re: 3. Yes, I know the test I'm suggesting is functionally the same as removing webhook.php. Either one works.

OK, so I downloaded a zip of your fork and put it in place of the current stripe extension. I then cleared the various Civi caches described [here](http://civicrm.stackexchange.com/questions/93/what-are-the-different-cache-clearing-mechanisms-relevant-when-developing) plus the Civi template cache. Then I made a recurring contribution (without having a webhook set in Stripe.php). The contribution_status_id in the civicrm_contribution table was 2 (aka Pending). Just to note, when making the contribution Civi did send me to the Thank You page and put the expected message:

```
Amount: $ 10.00 
Date: February 16th, 2016 4:57 PM
Transaction #: sub_7v73CqqdcAV94c
This recurring contribution will be automatically processed every month.

You will receive an email receipt which includes information about how to update or cancel this recurring contribution.
```

As expected, the subscription_id is in place of the transaction_id. But the contribution status is Pending. So can you disable the webhook and test whether the recurring contribution gets set to Completed on your end?

**Comment by h-c-c** [comment]
Re: 2.  Ok that makes sense, thanks.  I'll include this fix. 

Re: 3.  I'm not sure why you ask me if you know they're the same.  Give me some credit, OK?  I would never ask anyone specifically to perform a test I have not already performed myself.  

Wow, this is weird, I'm seeing a difference between how recurring contributions for memberships and those related to non-memberships are processed.  Without a webhook in place, the membership contribs show as complete and the non-membership contribs are pending, the way you see them.  Could you try the test with a recurring membership to confirm?  On our production site, all of are recurring contributions are exclusively for memberships, so that explains why everything is normal for us.  FWIW, I see the same behavior in 6.4.13.  

I thought maybe this difference was intrinsic to civi since there's apparently no code specific to memberships in Stripe.php. But since having tested the original fork and not seeing any difference between member/non-member recurring contribs status_id sans Webhook, I'm looking harder at my Stripe.php.

**Comment by h-c-c** [comment]
Mystery solved Re: 3.   It turns out if you a return a trxn_id in Stripe.php, a recurring membership contribution will be set to Completed for a valid transaction.  But a regular recurring contribution will be left Pending, requiring the aid of a webhook.  This inconsistency makes me think this is a bug.  If it were the same for both types of contributions, I'd think it would be more likely to be an undocumented feature.  I'll ask on IRC.

**Comment by drastik** [comment]
What is our status here?  Unfortunately I haven't had the time to dig deep on this.  The progress looks outstanding (nice work! :+1: ), but I see back & forth on `trxn_id, subscription_id, invoice_id` and other pieces.

**Comment by h-c-c** [comment]
Thank you!

It seems it could be "dealers choice" in terms of whether payproc developers decide to complete a successful recurring contribution in the core code, or with a webhook.  I intend to revert affa060 and go with the former because it acts on successful payment information the instant it is known, and removes one point of failure.   Granted, Stripe would have to be having a _really bad day_ if sent out invoice.payment_succeeded events with a significant delay. Apparently, the new way of doing this is returning "payment_status_id = 1" in Stripe.php. See:  https://issues.civicrm.org/jira/browse/CRM-16737  In any event, according to Eileen, "payment_status_id is the future!" 

Eileen thought the inconsistency with trxn_id sounded like a bug. FWIW, I confirmed it was the same behavior in 4.7 and opened a jira.   

In case you're interested...the instant way makes sense especially for my orgs use case...I built a door system for our makerspace that ties in with the civi database.  We're storing RFID card info in custom data and also checking membership status id.  When members join or re-up as recurring, it's nice and neighborly if their payment is reflected asap...so they can physically enter the space.  ;-)

The last commit is Boris's fix for a db query for subscription_id.  Thanks again, Boris!

**Comment by drastik** [comment]
Apologies for the late reply.
Your application sounds awesome @h-c-c .

I'd like to get all this work in, but we have merge conflicts now.  Are you still using a 4.6 branch or have you worked on top of this for your own installation(s)?

**Review by drastik — CHANGES_REQUESTED** [review]
Can you re-comment lines 355 - 359?

Good patch, thanks.

**Review by drastik — APPROVED** [review]
Thanks!

**Comment by drastik** [comment]
This thread got a bit crazy with all the patches.
What is still relevant?  I can't merge because of conflicts.

We need to re-roll against current (in a new Pull Request! ), just the pieces that should be updated.

**Comment by h-c-c** [comment]
Hey, sorry also for the late reply.  Thanks for the thanks!    Yeah, I guess I should be working on getting this merged first before new features/fixes.    I'll be in touch!

File: CRM/Core/Payment/Stripe.php [hunk 1]
```diff
@@ -352,11 +352,11 @@ function doDirectPayment(&$params) {
     */
 
     // drastik - Uncomment this for Drupal debugging to dblog.
-   /* 
+
```
Comment (drastik) [inline CRM/Core/Payment/Stripe.php]: We can't commit this un-commented, because `watchdog` is Drupal-specific (some people are on Wordpress / Joomla).

**Pull request closed without merging**